### PR TITLE
64-bit Taint Propagation Improvements

### PR DESCRIPTION
**Pull request by nathanjackson**
Initially, this work started out to support 64-bit multiply. Before, there was a warning message that would indicate that taint was not propagated through a 64-bit multiply due to the extension to 128-bits. However, we found multiple issues with 64-bit taint propagation and those issues are now fixed and I'll try to summarize everything going on here:

- Taint was not propagated through 64-bit multiply since the values were extended to 128-bit by TCG-LLVM. This issue was fixed by passing the low and hi bits of a 128-bit value to `taint_mul_compute`. To preserve the control bit mask calculations, we've updated the underlying data type to `llvm::APInt` which handles arbitrarily sized integers. We initially tried using the `unsigned __int128` type in GCC and Clang, but this was buggy.

- While testing the handling of 128-bit values, we discovered that the `cmpxchg16b` helper used 128-bit types, so we thought that this was a good candidate for testing helpers. It turned out that `cmpxchg16b` was not actually propagating taint due to an unhandled case where a host memory write to guest RAM was not considered in the host copy taint op. We fixed this issue by handling RAM pointers in this taint operation. This enables `cmpxchg16b` and others to propagate taint. With this change, you can expect some differences in your results, but we think they're correct.

- On the control bits side of things, we know there are some deficiencies that still have not been addressed (see: #371, #387). To test the 64-bit taint propagation, I had to implement some workarounds that would otherwise require significant re-work. There are a couple of issues here. First, `last_literal` is assumed to be the amount to shift by in shift instructions, so the assertion fails when built with `TAINT2_DEBUG` and LLVM is built with assertions enabled. However, this isn't the case and it is technically valid to shift more than the width of the value being shifted so I removed the assertion. Instead, we now explicitly handle this case by zeroing the control bits, which should be equivalent to the current behavior when assertions are disabled.

With these changes, you can expect some differences in taint results, but they seem to be reasonable. We've tested x86 and x86-64 fairly extensively and I feel confident in the results for these architectures. I do not have any testing in place for ARM and only the most basic tests for PPC (basically it doesn't crash).

### CODE REVIEW

File: panda/plugins/taint2/tests/update_cb_switch/update_cb_switch.cpp [hunk 1]
```diff
@@ -36,28 +40,32 @@
  *    size:  The number of bytes the LLVM instruction operates upon
  *    orig_cb_mask:  The original cb mask for the bytes being operated on
  *    orig_zero_mask:  The original zero mask for the bytes being operated on
+ *    wj/liters
```
Comment (AndrewFasano) [inline panda/plugins/taint2/tests/update_cb_switch/update_cb_switch.cpp:43]: Is this a typo?

Comment (nathanjackson) [reply panda/plugins/taint2/tests/update_cb_switch/update_cb_switch.cpp:43]: Yes. Will fix.

**Comment by nathanjackson** [comment]
It looks like rebasing fixes `make check`, converting to draft while I re-test.

**Comment by nathanjackson** [comment]
@AndrewFasano This one is ready to be merged\another review. My tests pass except for ARM replay (due to the snapshot issue) as discussed.

**Comment by jnmartin84** [comment]
sorry for junking up the "conversation" view of your PR with close/reopens and failures, forgot some dependencies on the Jenkins side

**Comment by nathanjackson** [comment]
@jnmartin84 No Problem, I understand that my PR is a guinea pig :hamster:

**Comment by jnmartin84** [comment]
panda was hung again so I killed the taint test

**Pull request merged**